src/pytypes/onefuzztypes/models.py

```diff
@@ -500,28 +506,43 @@ class WorkerEvent(EnumModel):
     running: Optional[WorkerRunningEvent]
 
 
-class SettingUpEventData(BaseModel):
+class NodeSettingUpEventData(BaseModel):
     tasks: List[UUID]
 
 
+class NodeDoneEventData(BaseModel):
+    error: Optional[str]
+    script_output: Optional[ProcessOutput]
+
+
+NodeStateData = Union[NodeSettingUpEventData, NodeDoneEventData]
+
+
 class NodeStateUpdate(BaseModel):
     state: NodeState
-    data: Optional[SettingUpEventData]
-
-    @validator("data")
-    def check_data(
-        cls,
-        data: Optional[SettingUpEventData],
-        values: Any,
-    ) -> Optional[SettingUpEventData]:
-        if data:
-            state = values.get("state")
-            if state and state != NodeState.setting_up:
-                raise ValueError(
-                    "data for node state update event does not match state = %s" % state
-                )
+    data: Optional[NodeStateData]
+
+    @root_validator(pre=False, skip_on_failure=True)
```

Comment:
I haven't actually tested this yet.

Reply:
We expect the `state_update` events to be adjacently tagged.

For now, `state_update.data` is optional, even for events which have a case in `NodeEventData`. This is for back compat.

In the future, we can make the `data` field mandatory for states that have data.

Examples:
Initial `free` start transition:
```json
{
    "state": "free"
}
```
Non-legacy enriched `setting_up` transition:
```json
{
    "state": "setting_up",
    "data": {
        "tasks": ["c6927d49-479d-4eda-a0f0-7084a31459da"]
    }
}
```
Legacy `done` transition, with no data:
```json
{
    "state": "done",
    "data": null
}
```